astropy/io/fits/_tiled_compression/tiled_compression.py

```diff
@@ -0,0 +1,576 @@
+            else:
+                zblank = None
+
+            if zblank is not None:
+                blank_mask = tile_data == zblank
+
+            if quantized:
+                dither_method = DITHER_METHODS[hdu._header.get("ZQUANTIZ", "NO_DITHER")]
+                dither_seed = hdu._header.get("ZDITHER0", 0)
+                q = Quantize(
+                    row=(irow + dither_seed) if dither_method != -1 else 0,
+                    dither_method=dither_method,
+                    quantize_level=None,
+                    bitpix=hdu._header["ZBITPIX"],
+                )
+                tile_data = np.asarray(
+                    q.decode_quantized(tile_data, row["ZSCALE"], row["ZZERO"])
+                ).reshape(actual_tile_shape)
+
+            if zblank is not None:
+                if not tile_data.flags.writeable:
+                    tile_data = tile_data.copy()
+                tile_data[blank_mask] = np.nan
+
+        data[tile_slices] = tile_data
+        istart[-1] += tile_shape[-1]
+        for idx in range(data.ndim - 1, 0, -1):
+            if istart[idx] >= data_shape[idx]:
+                istart[idx] = 0
+                istart[idx - 1] += tile_shape[idx - 1]
+
+    return data
+
+
+def compress_hdu(hdu):
+    """
+    Compress the data in a `~astropy.io.fits.CompImageHDU`.
+
+    The input HDU is expected to have a uncompressed numpy array as it's
+    ``.data`` attribute.
+
+    Parameters
+    ----------
+    hdu : `astropy.io.fits.CompImageHDU`
+        Input HDU to compress the data for.
+
+    Returns
+    -------
+    nbytes : `int`
+        The number of bytes of the heap.
+    heap : `bytes`
+        The bytes of the FITS table heap.
+    """
+
+    if not isinstance(hdu.data, np.ndarray):
+        raise TypeError("CompImageHDU.data must be a numpy.ndarray")
+
+    _check_compressed_header(hdu._header)
+
+    # TODO: This implementation is memory inefficient as it generates all the
+    # compressed bytes before forming them into the heap, leading to 2x the
+    # potential memory usage. Directly storing the compressed bytes into an
+    # expanding heap would fix this.
+
+    tile_shape = _tile_shape(hdu._header)
+    data_shape = _data_shape(hdu._header)
+
+    compressed_bytes = []
+    gzip_fallback = []
+    scales = []
+    zeros = []
+    zblank = None
+
+    irow = 0
+    istart = np.zeros(len(data_shape), dtype=int)
+
+    noisebit = _get_compression_setting(hdu._header, "noisebit", 0)
+
+    while True:
+
+        # In the following, we don't need to special case tiles near the edge
+        # as Numpy will automatically ignore parts of the slices that are out
+        # of bounds.
+        slices = tuple(
+            [
+                slice(istart[idx], istart[idx] + tile_shape[idx])
+                for idx in range(len(istart))
+            ]
+        )
+
+        data = hdu.data[slices]
+
+        settings = _header_to_settings(hdu._header, data.shape)
```

Comment:
this one could be moved outside of the loop ?

Reply:
We can't because it relies on data.shape which might change for edge tiles

Reply:
Oh right `data.shape` depends on the slices so it can change :+1: